Make sure all ba are defined in test, and only call Init once per PC

jmsexton03 opened the pull request:
## Summary

* Fixes a potential indexing issue in rr
* Makes ba[lev] definition use for loop instead of if to allow for nlevs>2
* Don't call ParticleInit functions on non-empty ParticleContainers

## Additional background
This is meant to fix #2243 
The assertions now depend on PR #2248 

## Checklist

The proposed changes:
- [X] fix a bug or incorrect behavior in AMReX
- [ ] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] are described in the proposed changes to the AMReX documentation, if appropriate

jmsexton03 committed:
Change boxes so all ba are defined, and use a new ParticleContainer for initializing tests

ax3l commented on Tests/Particles/GhostsAndVirtuals/main.cpp:
```diff
@@ -60,8 +60,9 @@ void test_ghosts_and_virtuals (TestParams& parms)
+    int n_fine = parms.nx;
     for (int lev = 1; lev < nlevs; lev++) {
-        int n_fine = parms.nx*rr[lev-1];
+        n_fine = rr[lev-1];
```
Comparing with previous value and below (`test_ghosts_and_virtuals_randomperbox`), should this read:
```suggestion
        n_fine *= rr[lev-1];
```
?

ax3l replied:
It's Friday afternoon, so all I can offer is pattern matching :D

jmsexton03 committed:
Remove duplicate workaround

ax3l commented on Tests/Particles/GhostsAndVirtuals/main.cpp:
```diff
@@ -40,7 +40,7 @@ void test_ghosts_and_virtuals (TestParams& parms)
-    Vector<int> rr(nlevs-1);
+    Vector<int> rr(nlevs);
     for (int lev = 1; lev < nlevs; lev++)
         rr[lev-1] = 2;
```
Since this is not performance-critical code, you can access with bounds-checked methods here:
```suggestion
        rr.at(lev-1) = 2;
```

ax3l commented on Tests/Particles/GhostsAndVirtuals/main.cpp:
```diff
@@ -60,8 +60,9 @@ void test_ghosts_and_virtuals (TestParams& parms)
+    int n_fine = parms.nx;
     for (int lev = 1; lev < nlevs; lev++) {
-        int n_fine = parms.nx*rr[lev-1];
+        n_fine *= rr[lev-1];
```
```suggestion
        n_fine *= rr.at(lev-1);
```

ax3l commented on Tests/Particles/GhostsAndVirtuals/main.cpp:
```diff
@@ -349,14 +350,15 @@ void test_ghosts_and_virtuals_randomperbox (TestParams& parms)
-        int n_fine = parms.nx*rr[0];
+    int n_fine = parms.nx;
+    for (int lev = 1; lev < nlevs; lev++) {
+        n_fine *= rr[lev-1];
```
```suggestion
        n_fine *= rr.at(lev-1);
```

ax3l commented on Tests/Particles/GhostsAndVirtuals/main.cpp:
```diff
@@ -329,7 +330,7 @@ void test_ghosts_and_virtuals_randomperbox (TestParams& parms)
-    Vector<int> rr(nlevs-1);
+    Vector<int> rr(nlevs);
     for (int lev = 1; lev < nlevs; lev++)
         rr[lev-1] = 2;
```
```suggestion
        rr.at(lev-1) = 2;
```

ax3l commented on Tests/Particles/GhostsAndVirtuals/main.cpp:
```diff
@@ -349,14 +350,15 @@ void test_ghosts_and_virtuals_randomperbox (TestParams& parms)
         // Build a box for the level 1 domain
         Box refined_patch(refined_lo, refined_hi);
-        ba[1].define(refined_patch);
+        ba[lev].define(refined_patch);
```
```suggestion
        ba.a(lev).define(refined_patch);
```

ax3l commented on Tests/Particles/GhostsAndVirtuals/main.cpp:
```diff
@@ -479,7 +486,7 @@ void test_ghosts_and_virtuals_onepercell (TestParams& parms)
-    Vector<int> rr(nlevs-1);
+    Vector<int> rr(nlevs);
     for (int lev = 1; lev < nlevs; lev++)
         rr[lev-1] = 2;
```
```suggestion
        rr.at(lev-1) = 2;
```

ax3l reviewed (commented):
LGTM, thank you.
Little comment on `std::vector::at` that can catch this: https://en.cppreference.com/w/cpp/container/vector/at

Will push a PR to your branch to enable CI.

ax3l committed:
CI: Test Particles_GhostsAndVirtuals

ax3l committed:
GhostsAndVirtuals: Cleanup

- Safety: use `.at()` over `operator[]` for performance-uncritical
  code
- Some formatting

WeiqunZhang commented:
Vector::operator[] can catch out of bound error when assertion or debug is
on.

On Mon, Aug 16, 2021, 5:57 PM Axel Huebl ***@***.***> wrote:

> ***@***.**** commented on this pull request.
>
> LGTM, thank you.
> Little comment on std::vector::at that can catch this:
> https://en.cppreference.com/w/cpp/container/vector/at
>
> Will push a PR to your branch to enable CI.
> ------------------------------
>
> In Tests/Particles/GhostsAndVirtuals/main.cpp
> <https://github.com/AMReX-Codes/amrex/pull/2247#discussion_r689954472>:
>
> > @@ -40,7 +40,7 @@ void test_ghosts_and_virtuals (TestParams& parms)
>
>      const Box domain(domain_lo, domain_hi);
>
>
>
>      // Define the refinement ratio
>
> -    Vector<int> rr(nlevs-1);
>
> +    Vector<int> rr(nlevs);
>
>      for (int lev = 1; lev < nlevs; lev++)
>
>          rr[lev-1] = 2;
>
>
> Since this is not performance-critical code, you can access with
> bounds-checked methods here:
> ⬇️ Suggested change
>
> -        rr[lev-1] = 2;
>
> +        rr.at(lev-1) = 2;
>
>
> ------------------------------
>
> In Tests/Particles/GhostsAndVirtuals/main.cpp
> <https://github.com/AMReX-Codes/amrex/pull/2247#discussion_r689954619>:
>
> >      for (int lev = 1; lev < nlevs; lev++) {
>
> -        int n_fine = parms.nx*rr[lev-1];
>
> +        n_fine *= rr[lev-1];
>
>
> ⬇️ Suggested change
>
> -        n_fine *= rr[lev-1];
>
> +        n_fine *= rr.at(lev-1);
>
>
> ------------------------------
>
> In Tests/Particles/GhostsAndVirtuals/main.cpp
> <https://github.com/AMReX-Codes/amrex/pull/2247#discussion_r689954778>:
>
> > @@ -349,14 +350,15 @@ void test_ghosts_and_virtuals_randomperbox (TestParams& parms)
>
>      Vector<BoxArray> ba(nlevs);
>
>      ba[0].define(domain);
>
>
>
> -    if (nlevs > 1) {
>
> -        int n_fine = parms.nx*rr[0];
>
> +    int n_fine = parms.nx;
>
> +    for (int lev = 1; lev < nlevs; lev++) {
>
> +        n_fine *= rr[lev-1];
>
>
> ⬇️ Suggested change
>
> -        n_fine *= rr[lev-1];
>
> +        n_fine *= rr.at(lev-1);
>
>
> ------------------------------
>
> In Tests/Particles/GhostsAndVirtuals/main.cpp
> <https://github.com/AMReX-Codes/amrex/pull/2247#discussion_r689954931>:
>
> > @@ -329,7 +330,7 @@ void test_ghosts_and_virtuals_randomperbox (TestParams& parms)
>
>      const Box domain(domain_lo, domain_hi);
>
>
>
>      // Define the refinement ratio
>
> -    Vector<int> rr(nlevs-1);
>
> +    Vector<int> rr(nlevs);
>
>      for (int lev = 1; lev < nlevs; lev++)
>
>          rr[lev-1] = 2;
>
>
> ⬇️ Suggested change
>
> -        rr[lev-1] = 2;
>
> +        rr.at(lev-1) = 2;
>
>
> ------------------------------
>
> In Tests/Particles/GhostsAndVirtuals/main.cpp
> <https://github.com/AMReX-Codes/amrex/pull/2247#discussion_r689955287>:
>
> >          IntVect refined_lo(AMREX_D_DECL(n_fine/4,n_fine/4,n_fine/4));
>
>          IntVect refined_hi(AMREX_D_DECL(3*n_fine/4-1,3*n_fine/4-1,3*n_fine/4-1));
>
>
>
>          // Build a box for the level 1 domain
>
>          Box refined_patch(refined_lo, refined_hi);
>
> -        ba[1].define(refined_patch);
>
> +        ba[lev].define(refined_patch);
>
>
> ⬇️ Suggested change
>
> -        ba[lev].define(refined_patch);
>
> +        ba.a(lev).define(refined_patch);
>
>
> ------------------------------
>
> In Tests/Particles/GhostsAndVirtuals/main.cpp
> <https://github.com/AMReX-Codes/amrex/pull/2247#discussion_r689955445>:
>
> > @@ -479,7 +486,7 @@ void test_ghosts_and_virtuals_onepercell (TestParams& parms)
>
>      const Box domain(domain_lo, domain_hi);
>
>
>
>      // Define the refinement ratio
>
> -    Vector<int> rr(nlevs-1);
>
> +    Vector<int> rr(nlevs);
>
>      for (int lev = 1; lev < nlevs; lev++)
>
>          rr[lev-1] = 2;
>
>
> ⬇️ Suggested change
>
> -        rr[lev-1] = 2;
>
> +        rr.at(lev-1) = 2;
>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/AMReX-Codes/amrex/pull/2247#pullrequestreview-731244243>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB37TYJC7Z6TOJWRIM4PN7TT5GXWFANCNFSM5CEQHMTA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

ax3l commented:
That seems to be GNU libstdc++ specific though: https://stackoverflow.com/a/48804020/2719194

Since the code parts are not in hot loops here, I would exercise the `.at()` that would have caught the original UB issue with any C++ implementation.

WeiqunZhang commented:
Oh I was talking about amrex::Vector.

On Mon, Aug 16, 2021, 6:28 PM Axel Huebl ***@***.***> wrote:

> That seems to be GNU libstdc++ specific though:
> https://stackoverflow.com/a/48804020/2719194
>
> Since the code parts are not in hot loops here, I would exercise the .at()
> that would have caught the original issue with any C++ implementation.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/AMReX-Codes/amrex/pull/2247#issuecomment-899924820>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB37TYOL7CUCDL7AUESDCSLT5G3LFANCNFSM5CEQHMTA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

ax3l commented:
Oh yes, you overwrite `operator[]` after deriving from `std::vector` - smart.
